src/main/cpp/src/datetime_utils.cuh

```diff
@@ -23,91 +23,213 @@
+ * The formula is: Int.MaxValue/days_per_year + 1970.
+ */
+struct date_segments {
+  /**
+   * @brief Constructor a default date segments.
+   * By default, use epoch date: "1970-01-01".
+   */
+  __device__ date_segments() : year(1970), month(1), day(1) {}
+
+  __device__ bool is_valid_date() const
   {
-    if (month == 2) { return is_leap_year() ? 29 : 28; }
-    return (month == 4 || month == 6 || month == 9 || month == 11) ? 30 : 31;
+    return date_time_utils::is_valid_date_for_date(year, month, day);
   }
 
-  __device__ bool is_valid_date() const
+  /**
+   * @brief Get days since epoch 1970-01-01.
+   * Can handle all int years.
+   */
+  __device__ int64_t to_epoch_day() const
   {
-    // rough check for year range
-    if (year < -300000 || year > 300000) { return false; }
+    return date_time_utils::to_epoch_day(year, month, day);
+  }
 
-    if (month < 1 || month > 12 || day < 1) {
-      return false;  // Invalid month or day
-    }
+  int32_t year;
 
-    // Check for leap year, February has 29 days in leap year
-    if (month == 2 && is_leap_year()) { return (day <= 29); }
+  // 1-12
+  int32_t month;
 
-    // Check against the standard days
-    return (day <= days_in_month());
+  // 1-31; it is 29 for leap February, or 28 for regular February
+  int32_t day;
+};
+
+/**
+ * @brief Represents local date time in a timezone with microsecond accuracy.
+ * Spark stores timestamp into Long in microseconds.
+ * A Long is able to represent a timestamp with max 6 digits of microseconds.
+ * The formula is: Long.MaxValue/microseconds_per_year + 1970.
+ */
+struct ts_segments {
+  /**
+   * @brief Constructor a default timestamp segments.
+   * By default, use epoch date with mid-night time: "1970-01-01 00:00:00.000000".
+   */
+  __device__ ts_segments()
+    : year(1970), month(1), day(1), hour(0), minute(0), second(0), microseconds(0)
+  {
+  }
+
+  /**
+   * @brief Is this timestamp segments valid.
+   */
+  __device__ bool is_valid_ts() const
+  {
+    return date_time_utils::is_valid_date_for_timestamp(year, month, day) &&
+           date_time_utils::is_valid_time(hour, minute, second, microseconds);
   }
 
-  __device__ bool is_valid_time() const
+  /**
+   * @brief Get days since epoch 1970-01-01.
+   * Can handle all int years.
+   */
+  __device__ int64_t to_epoch_day() const
   {
-    return (hour >= 0 && hour < 24) && (minute >= 0 && minute < 60) &&
-           (second >= 0 && second < 60) && (microseconds >= 0 && microseconds < 1000000);
+    return date_time_utils::to_epoch_day(year, month, day);
   }
+
+  // max 6 digits for Spark timestamp
+  int32_t year;
+
+  // 1-12
+  int32_t month;
+
+  // 1-31; it is 29 for leap February, or 28 for regular February
+  int32_t day;
+
+  // 0-23
+  int32_t hour;
+
+  // 0-59
+  int32_t minute;
+
+  // 0-59
+  int32_t second;
+
+  // 0-999999, only parse 6 digits, ignore/truncate the rest digits
+  int32_t microseconds;
```

Comment:
We may not need these all the time. How about separating them into a new/derived class for saving memory when they are not needed? In particular:
```
struct date_segment {
int year, month, day;
};

struct ts_segment: date_segment {
int hour, minute, second, microsecond;
};
```
So when we process only date, we won't have to allocate extra memory for the hour/minutes etc stuff that are never touched.

Reply:
You are right, could it be a follow-up? And I would use composite types instead of inherited types.

```cpp
struct date_segments {
  int year, month, day;
};

struct time_segments {
  int hour, minute, second, microsecond;
}

struct timestamp_segments {
  date_segment date_segments;
  time_segment time_segments;
}

parse_int(ts.year) => parse_int(ts.date_segments.year)

```cpp